### ISSUE TITLE
chore(typing): adopt jaxtyping for precise shape/dtype annotations

### ISSUE THREAD

**Issue by antonioterpin** [antonioterpin/pinet#97]
## Motivation

Across `src/pinet/` we annotate tensors as `jnp.ndarray` / `jax.Array` and repeat the shape and dtype in every docstring. This is both imprecise (the type checker cannot catch a wrong-rank argument) and unmaintainable (shape comments drift out of sync with the signature). [jaxtyping](https://docs.kidger.site/jaxtyping/) lets us encode rank, dtype and symbolic axis sizes in the annotation itself, so the signature becomes the source of truth and docstrings no longer need to carry shape/dtype at all.

Concrete examples where this bites us today:

- [src/pinet/dataclasses.py:39-41](src/pinet/dataclasses.py#L39-L41) — `EqualityConstraintsSpecification` has `b`, `a_dyn`, `a_dyn_pinv` all typed `jnp.ndarray`, with the shape `(batch_size, n_constraints, dimension)` encoded only in the docstring.
- [src/pinet/constraints/affine_equality.py:30-40](src/pinet/constraints/affine_equality.py#L30-L40) — `EqualityConstraint.__init__` takes `a_dyn` and `b` as bare `jnp.ndarray`, then `setup()` re-checks rank/shape at runtime with asserts that jaxtyping would catch at trace time.
- [src/pinet/project.py:386-438](src/pinet/project.py#L386-L438) — `_project_general` has six tensor-shaped arguments, all `jnp.ndarray`, with scaling factors `d_r` (rows) and `d_c` (columns) distinguished only by naming.
- [src/pinet/solver/admm.py:82-143](src/pinet/solver/admm.py#L82-L143) — `build_iteration_step` builds an iteration that slices `sk.x[:, :dim, :]` vs `sk.x[:, dim:, :]`; the lifted-vs-original dimension split is today a comment.
- [src/pinet/equilibration.py:10-12](src/pinet/equilibration.py#L10-L12) — `ruiz_equilibration` signature hides the `a_dyn` is 2D vs the batched 3D `a_dyn` used everywhere else (this is currently enforced by an assert).

## Proposal

1. Add `jaxtyping` as a runtime dependency (`uv add jaxtyping`).
2. Replace `jnp.ndarray` / `jax.Array` annotations with dtype-parameterised jaxtyping aliases using symbolic axis names. Example:
   ```python
   from jaxtyping import Array, Float, Bool
   BatchedMat = Float[Array, \"B m n\"]
   BatchedVec = Float[Array, \"B m 1\"]
   Mask       = Bool[Array, \"d\"]
   ```
   Symbolic names (`B`, `m`, `n`, `d`, `d_lifted`, …) are resolved at first call and consistency-checked across call sites of annotated functions, which is exactly the invariant we are asserting by hand today.
3. Strip shape/dtype sentences from docstrings once the signature carries them. Per [docs/standards/typing-docstrings.md](docs/standards/typing-docstrings.md), type info belongs in signatures — this just extends the rule to shapes.
4. Adjust ruff config in [pyproject.toml](pyproject.toml) to ignore `F722` (syntax error in forward annotation) and `F821` (undefined name) inside jaxtyping-annotated functions, since the string annotations \"B m n\" trip both rules. Prefer a targeted `per-file-ignores` or the `typing` extension over a project-wide disable.

## Suggested rollout

- [ ] Introduce a small `pinet/_typing.py` with shared aliases (`EqMat`, `BatchedVec`, `Mask`, `Scaling`, …) so call-site annotations stay short.
- [ ] Migrate the dataclasses in [src/pinet/dataclasses.py](src/pinet/dataclasses.py) first (highest leverage — they flow everywhere).
- [ ] Migrate `Constraint` subclasses ([affine_equality.py](src/pinet/constraints/affine_equality.py), [affine_inequality.py](src/pinet/constraints/affine_inequality.py), [box.py](src/pinet/constraints/box.py), [soc_constraint.py](src/pinet/constraints/soc_constraint.py), [cartesian_constraint.py](src/pinet/constraints/cartesian_constraint.py), [non_linear.py](src/pinet/constraints/non_linear.py)).
- [ ] Migrate the solver surface ([solver/admm.py](src/pinet/solver/admm.py), [project.py](src/pinet/project.py), [equilibration.py](src/pinet/equilibration.py)).
- [ ] Delete shape/dtype sentences from docstrings that the signature now carries.

Runtime shape/dtype enforcement (typeguard/beartype wiring) is tracked separately to keep this PR scoped to annotations.

**Comment by antonioterpin** [comment]
Landed via #104 (jaxtyping aliases) + #117 (tools/benchmarks rollout).
